### PR TITLE
plumed: cxx11 portgroup

### PR DESCRIPTION
**Pull request by GiovanniBussi**
I included cxx11 PortGroup (see hint from @kencu at #1252). Compilers from gcc family will be disabled now. I took the occasion to update the snapshot used in `plumed-devel` subport (development branch).

### CODE REVIEW

**Comment by kencu** [comment]
unfortunately that didn't work. build fails on buildbots due to what appears to be header differences. Sadly this looks to be in the 1% where this PG doesn't "just work". May need a more elegant fix...

**Comment by GiovanniBussi** [comment]
@kencu should I revert the change? I am not sure it makes sense to apply the PortGroup if it's not working correctly. Could enforcing clang-38 on old operating systems (as I proposed in #1252) be a better idea? Clearly, the two changes to not exclude each other.

**Comment by kencu** [comment]
Let me look into it a bit, if you don't mind. For now, it still works on all the systems it used to work on, so nobody has lost any functionality.

Making it build with clang against the gcc7 headers is what is causing the new issue, I believe. Some ```#define``` is missing, more than likely, on some gcc-specific header guard, is often the case. 

You said you have a 10.6.8 machine set up. Did this build on that machine with the PortGroup added? 

For example, you mentioned you needed some c++11 features, so maybe this is one of the ports where we just need to blacklist older clang compilers.

**Comment by GiovanniBussi** [comment]
OK I'll try and let you know, thanks

**Comment by GiovanniBussi** [comment]
I made the following tests:

- Before syncing port repository (i.e. without PortGroup cxx11):
  - `port install plumed +gcc7` compiles correctly.
  - `port install plumed +clang50` fails, not finding `<initializer_list>` include file (c++11).
- After syncing port repository (i.e. with PortGroup cxx11):
  - `port install plumed +gcc7` compiles correctly.
  - `port install plumed +clang50` fails at the link stage

Let me know how you think it is better to proceed. Thanks!

**Comment by kencu** [comment]
Lets see what we know:

- we know all builds against libc++ with any modern clang (3.9 or greater, say) are likely OK. Might have to test exactly how new the clang needs to be.
- I'm not sure if any builds with clang on systems < 10.9 are working. You said in #1252 that clang-3.8 built it correctly on your 10.6.8 laptop. Not sure I understand why that would work, yet clang-5.0 would not, but could be...
- gcc builds on systems < 10.9 work, but link against gcc7's libstdc++, which works out OK so long as the system is set up to use gcc7's libstdc++ (ie the cxx11 1.1 PortGroup). This would not work correctly on systems configured for libc++, though, so they would need to be excluded from building it with gcc7 (and they don't appear to need to build it with gcc7).

So we could set up a test for libc++, and force gcc7 on systems < 10.9 that are not using libc++. That would probably work out.

Or we can figure out why the clang builds using the gcc7 headers are failing ( I like this better, but it's way harder to do...).

**Comment by kencu** [comment]
So this hashtable error building against the gcc headers with clang (or sometimes with other gcc versions or with the intel compiler) is pretty easy to find when googling -- lots of references to it. The solution always seems to be "don't do that", with no actual fix listed anywhere that I have been able to find. Perhaps @MarcusCalhoun-Lopez knows how to get this to work correctly.

If not, then I guess the best road forward would probably be to so something like this in the Portfile pseudocode:
```
if (building with configure.cxx_stdlib eq libc++)
    make sure the compiler is a new enough clang, otherwise leave it alone
else
    force gcc7 on < 10.9
endif
```

**Comment by kencu** [comment]
I'll ping @RJVB on this as well, as he very frequently uses clang against the gcc headers I believe, and has no doubt come across this hashtable error as well.

**Comment by RJVB** [comment]
Erm, what hashtable error?

I do recognise the missing <initializer_list> problem from when I was still running 10.6 and even have a vague memory how I worked around it (build with a new enough GCC, or patch the code).
Patching the code is probably no longer an option.

BTW:
> port install plumed +clang50 fails, not finding <initializer_list> include file (c++11).

is it at least correct, that is, is the file really missing?

> he very frequently uses clang against the gcc headers

Not really. What I do *sometimes* is build using G++ against the clang headers (and libc++).

Yeah, on Linux... It's a thought though if on systems that haven't been converted to using libc++. If GCC builds a given project, chances are that clang with `-stdlib=libstdc++` will do so too provided it uses the same libstdc++ and corresponding headers. Of course the latest clang version I ever did something like that with on Mac isn't very recent, probably 3.4 .

**Comment by kencu** [comment]
Here's a link to a typical build failure, when clang-5.0 using the c++11 portgroup tries to build this port against the gcc7 headers <https://build.macports.org/builders/ports-10.8_x86_64_legacy-builder/builds/50179/steps/install-port/logs/stdio>

It can't find symbols to
```
std::__detail::_Prime_rehash_policy::_M_need_rehash
```
and others

**Comment by RJVB** [comment]
On Wednesday January 31 2018 12:07:05 Ken wrote:


>It can't find symbols to
>```
>std::__detail::_Prime_rehash_policy::_M_need_rehash
>```

The link command doesn't show which C++ runtime is being used. I suppose you have check (run the link command locally with the -v option)?

Does this code build with clang on Linux?

R.

**Comment by GiovanniBussi** [comment]
Dear all,

thanks for your inputs. I actually double checked and the information that I was able to compile on 10.6 with clang 3.8 was incorrect. The only thing that works there is gcc. Sorry but I don't have always access to that machine, plus it is really really slow (the first intel MacBook, from 2006).

So @kencu the solution that you proposed is this one:

    if (building with configure.cxx_stdlib eq libc++)
        make sure the compiler is a new enough clang, otherwise leave it alone
    else
       force gcc7 on < 10.9
    endif

Do you mean that I should add this to my Portfile or is it something that should go in the cxx11 PortGroup?

Thanks again!

Giovanni

**Comment by kencu** [comment]
> The link command doesn't show which C++ runtime is being used.

Oh, good pickup! You'd think I'd have noticed that, as I fix that error all day long. Perhaps that is what we need to do here. Spec the standard c++ library in the ld flags. That's certainly easy to add. I'll try that when I can get back to that machine.

>Do you mean that I should add this to my Portfile or is it something that should go in the cxx11 PortGroup?

Neither, really. 

If nobody can come up with a way to fix the hashtable errors, we'll need to write up some proper tcl code that does what that pseudocode does, and put it in your portfile.

**Comment by GiovanniBussi** [comment]
Yes I was planning to convert it to TCL first :-) I was just wondering if there are other C++11 codes around they might benefit from this. Or it could be an option of the cxx11 PortGroup.

Anyway, fine for me adding it to my port. I will try to do it.

Thanks again

Giovanni

**Comment by kencu** [comment]
actually, if you don't mind, try adding something like this to your Portfile instead:
```
configure.ldflags-append ${configure.cxx_stdlib}
```

**Comment by kencu** [comment]
Try the "fixed" version perhaps that I posted above, or wait and I'll try it when I can get back to the machine I have. Always hard to "wing it" on the fly without a proper setup to test it on. Thanks!

**Comment by kencu** [comment]
OK. So confirmed that explicitly adding the proper stdlib to the link line fixes the build. That is great, and the proper fix for now. Thanks, @RJVB ! I can see why that happens.

Now ideally, this should not be needed at all. Somewhere in the build scripts for plumed, the CXX flags should be added onto the LD flags, and then this would be taken care of.

In the absence of that, we can add it explicitly. One final hiccup -- if we add that ldflag onto a gcc build, it will fail as gcc presently has no idea what to do with the -stdlib flag, and chokes on it.

```
if [string match *clang* ${configure.compiler}] {
    configure.ldflags-append -stdlib=${configure.cxx_stdlib}
}
```

If we put together a PR that does that, and squash it all into one commit, we should be good to go! AFAICT, all systems should build it then without any troubles. (Haven't tested it on PPC, of course).

**Comment by RJVB** [comment]
> Now ideally, this should not be needed at all. Somewhere in the build scripts for plumed, the CXX flags should be added onto the LD flags, and then this would be taken care of.

I'd go one step further: on "legacy" systems clang ought to be confirmed to use -stdlib=libstdc++ as its built-in default.

**Comment by kencu** [comment]
Oh, and one more piece of good news. If you wrap that block above in a 
```
post-configure {

}
```
block, I think you'll find your gcc variants all work again.

**Comment by kencu** [comment]
>'d go one step further: on "legacy" systems clang ought to be confirmed to use -stdlib=libstdc++ as its built-in default.

Actually, that's the problem here. With the cxx11 1.1 portgroup, you want clang to use -stdlib=macports-libstdc++. But if there is no -stdlib spec'd on the link line, it defaults to -stdlib=libstdc++, and that one has no c++11 support, and dies.

**Comment by GiovanniBussi** [comment]
@kencu thanks a lot!

So, you propose a pull request like this:
````
post-configure {
    if [string match *clang* ${configure.compiler}] {
        configure.ldflags-append -stdlib=${configure.cxx_stdlib}
    }
}
````
This confuses me, and seems a bit illogical. Why should I modify LDFLAGS *after* `./configure`? Isn't it clearer without the `post-configure` statement? By the way, I tried the version without the `post-configure` statement and it looks it works. If you think that's ok I will open a request with that one.

The reason why I would not like to add CXXFLAGS to LDFLAGS is rather complicated, I write it below anyway. Maybe you have alternative suggestions. ***Ignore the text below if it's too long***.

PLUMED provides a C++ library that should be linkable using other compilers (including Fortran). In particular, it should be combined with other software that typically makes the final linking using the Fortran compiler, and the change to the link command in order to load PLUMED should be not too intrusive. The only portable way we found is to do something like
````
$(FC) fortran_obj.o $(PLUMED_LOAD) -o static_executable
````
where `$(PLUMED_LOAD)` is
- `/usr/lib/libplumed.so` for dynamic binding
- `/usr/lib/plumed/some-object.o $(LDFLAGS)` for static binding (required on some machine)
During our `./configure` script, in order to allow this behavior we test if adding e.g. `-lstdc++`  or `-lc++` allow for a correct linking and, if necessary, we add it to `$(LDFLAGS)`. Notice that this part is skipped on OSX since we always had trouble with static linking here, plus it is not necessary (dylib always works).

Bottom line: if I add `$(CXXFLAGS)` to `$(LDFLAGS)` I will risk to pass options that are meant for a C++ compiler to a C or Fortran compiler. That's why I would prefer to keep them separate, allowing the final user to customize them when calling `./configure`. Notice that we always combine `$(LIBS)` and `$(LDFLAGS)`, maybe that's the mistake.

Thanks again!

Giovanni

**Comment by RJVB** [comment]
>Actually, that's the problem here. With the cxx11 1.1 portgroup, you want clang to use -stdlib=macports-libstdc++. But if there is no -stdlib spec'd on the link line, it defaults to -stdlib=libstdc++, and that one has no c++11 support, and dies.

>>'d go one step further: on "legacy" systems clang ought to be confirmed to use -stdlib=macports-libstdc++ as its built-in default.

T,FTFY ;)

**Comment by RJVB** [comment]
>Bottom line: if I add `$(CXXFLAGS)` to `$(LDFLAGS)` I will risk to pass options that are meant for a C++ compiler to a C or Fortran compiler. That's why I would prefer to keep them separate, allowing the final user to customize them when calling `./configure`. Notice that we always combine `$(LIBS)` and `$(LDFLAGS)`, maybe that's the mistake.

In that case, check what configure.cxx_stdlib actually contains, and use (rewrite) it to a form that any compiler will accept. The target flag here will ultimately be something like -lc++ or -lmacports-stdc++, and any compiler should accept that in a link command. But if necessary you can probably wrap it with -Wl to remove any ambiguity.

There may also be the alternative to specify "LDFLAGS+=-lfoo" on the make command line (build.args or build.post_args).

**Comment by kencu** [comment]
>By the way, I tried the version without the post-configure statement and it looks it works

Try it with the gcc variants. You will probably find, as I did, that only in a post-configure block (or pre-build block, if you like that better) is the configure.compiler set correctly to make the correct decision about adding the stdlib flag.

>we test if adding e.g. -lstdc++ or -lc++ allow for a correct linking and, if necessary, we add it to $(LDFLAGS). Notice that this part is skipped on OSX

By not specifying the CXX flags to be added to the LD flags, you leave the `-stdlib` unset during linking on MacOS. That kicks in default settings, which usually will work, but not when we do fancy things like allowing c++11 on older systems. In that case, MacPorts needs to override default behaviours of clang by specifying the full CXX and LD flags. 

It's fine -- you can add it manually in the Portfile if that's your preference. 

I don't think homebrew has anything like the cxx11 PortGroup functionality, so it "just won't work" on older systems there. Or it can build against gcc's libstdc++ and you take your chances with breakage.  

Kudos to us, we have Jeremy, Marcus, and Ryan to make all this stuff work.

**Comment by kencu** [comment]
the gcc build variants only show up in systems  < 10.9 that are not configured for libc++, but this is OK because that is the only place those gcc builds are safe to use anyway.

If you don't use the post-configure block, sooner or later someone will try to build plumed with a gcc variant on one of those systems, and it will error out, and then there will be tickets about it.

**Comment by GiovanniBussi** [comment]
Good point... So I will use the whole block. Thanks again to everyone, I
actually learned a lot with this discussion!

Giovanni

**Comment by kencu** [comment]
Drat. I must have got something wrong. Sorry. I'll take it out of the post-configure block and re-think how and if we can still enable those gcc variants for you.

**Comment by GiovanniBussi** [comment]
Hi,

for me it's no problem to disable gcc. It is disabled on new OS, and it
will be disabled in old ones.

Giovanni
